runtime/lua/provider/python/health.lua

```diff
@@ -0,0 +1,490 @@
+        end
+      end
+    end
+  end
+
+  if not is_blank(python_exe) and not vim.g[host_prog_var] then
+    if
+        is_blank(venv)
+        and not is_blank(pyenv)
+        and not is_blank(pyenv_root)
+        and vim.startswith(vim.fn.resolve(python_exe), pyenv_root .. '/')
+    then
+      local advice = string.format(
+        'Create a virtualenv specifically for Nvim using pyenv, and set `g:%s`.  This will avoid the need to install the pynvim module in each version/virtualenv.',
+        host_prog_var)
+      health.warn('pyenv is not set up optimally.', advice)
+    elseif not is_blank(venv) then
+      local venv_root
+      if not is_blank(pyenv_root) then
+        venv_root = pyenv_root
+      else
+        venv_root = vim.fs.dirname(venv)
+      end
+
+      if vim.startswith(vim.fn.resolve(python_exe), venv_root .. '/') then
+        local advice = string.format(
+          'Create a virtualenv specifically for Nvim and use `g:%s`.  This will avoid the need to install the pynvim module in each virtualenv.',
+          host_prog_var)
+        health.warn('Your virtualenv is not set up optimally.', advice)
+      end
+    end
+  end
+
+  if is_blank(python_exe) and not is_blank(pyname) then
+    -- An error message should have already printed.
+    health.error('`' .. pyname .. '` was not found.')
+  elseif not is_blank(python_exe) and not check_bin(python_exe) then
+    python_exe = ''
+  end
+
+  -- Diagnostic output
+  health.info('Executable: ' .. (is_blank(python_exe) and 'Not found' or python_exe))
+  if vim.tbl_count(python_multiple) > 0 then
+    for _, path_bin in ipairs(python_multiple) do
+      health.info('Other python executable: ' .. path_bin)
+    end
+  end
+
+  if is_blank(python_exe) then
+    -- No Python executable can import 'neovim'. Check if any Python executable
+    -- can import 'pynvim'. If so, that Python failed to import 'neovim' as
+    -- well, which is most probably due to a failed pip upgrade:
+    -- https://github.com/neovim/neovim/wiki/Following-HEAD#20181118
+    local pynvim_table = vim.fn['provider#pythonx#DetectByModule']('pynvim', 3)
+    local pynvim_exe = pynvim_table[1]
+    if not is_blank(pynvim_exe) then
+      local message = 'Detected pip upgrade failure: Python executable can import "pynvim" but not "neovim": '
+          .. pynvim_exe
+      local advice = {
+        'Use that Python version to reinstall "pynvim" and optionally "neovim".',
+        pynvim_exe .. ' -m pip uninstall pynvim neovim',
+        pynvim_exe .. ' -m pip install pynvim',
+        pynvim_exe .. ' -m pip install neovim  # only if needed by third-party software',
+      }
+      health.error(message, advice)
+    end
+  else
+    local version_info_table = version_info(python_exe)
+    local pyversion = version_info_table[1]
+    local current = version_info_table[2]
+    local latest = version_info_table[3]
+    local status = version_info_table[4]
+
+    if not vim.version.range('~3'):has(pyversion) then
```

Comment:
Could we not check if version is less than 3?

```suggestion
    if vim.version.lt(pyversion, 3) then
```